torch/csrc/jit/codegen/cuda/lower_utils.cpp

```diff
@@ -304,6 +311,55 @@ std::unordered_map<ParallelType, IterDomain*, TypeHash> getParallelDomains(
   return parallel_domains;
 }
 
+bool isCpAsyncInit(const Expr* expr) {
+  return isTensorScalarFillOp(expr) &&
+      // FIXME:
+      //  We'd need to add a flag to all the init
+      //   exprs so we could robustly detect initialization
+      //   in all cases.
+      isCpAsyncOp(getTvOutput(expr)->definition());
+}
+
+c10::optional<Expr*> getMaybePredicatedSingleton(Expr* expr) {
+  if (auto ite = dynamic_cast<kir::IfThenElse*>(expr)) {
+    if (ite->elseBody().empty()) {
+      if (ite->thenBody().size() == 1) {
+        return ite->thenBody().exprs()[0];
+      }
+    }
+  }
+  return c10::nullopt;
+}
+
+//! Short-cut for checking if the expression loads from global memory.
+bool isGlobalLoad(const Expr* expr) {
+  if (expr->isA<LoadStoreOp>() ||
+      (expr->isA<UnaryOp>() &&
+       expr->as<UnaryOp>()->getUnaryOpType() == UnaryOpType::Set)) {
+    if (auto in_tv = dynamic_cast<TensorView*>(expr->input(0))) {
```

Comment:
```suggestion
    if (auto in_tv = getTv(expr->input(0))) {
```

This would work even when `expr->input(0)` is a `kir::TensorIndex`.